cdap-watchdog/src/main/java/io/cdap/cdap/metrics/query/MetricsQueryHelper.java

```diff
@@ -280,14 +280,33 @@ private void setTimeRangeInQueryRequest(MetricQueryRequest request, Map<String,
     }
   }
 
-  private Integer getResolution(String resolution, Long start, Long end) {
-    if (resolution.equals(PARAM_AUTO_RESOLUTION)) {
+  /**
+   * Get the integer resolution based on the resolution string, start and end ts. The logic of determining resolution
+   * is:
+   * 1. If the resolution string is a specific time interval, for example, 1s, 1m, 60s, and if this interval exists in
+   * the available resolution, that resolution will get returned.
+   * 2. If the resolution "auto", then start and end timestamp must be specified to determine the correct resolution.
+   * If end - start > 10 hours, hour resolution will be used. If end - start > 10 mins, min resolution will be used.
+   * If end -start < 10 mins, minimum resolution will be used.
+   * 3. If the resolution is null, i.e, not specified in the query, then if start and end timestamp are not null,
+   * the logic will be same as the resolution is "auto". If any of the start or end timestamp is not specified,
+   * minimum resolution will be used.
+   *
+   * @param resolution the resolution string, can be specific resolution like 1s, 1m, etc, or can be auto or null.
+   * @param start the start timestamp, null if not specified in the query
+   * @param end the end timestamp, null if not specified in the query
+   * @return the integer resolution for the query
+   */
+  private Integer getResolution(@Nullable String resolution, @Nullable Long start, @Nullable Long end) {
```

Comment:
Actually, this method should have a unit test. The end2end test you have is great but it is better to test at the lowest level, that allows testing a lot more edge cases.

Reply:
👍 I wanted to add unit test for this class initially but essentially found there is not a single unit test for this class. And this method is just a small method in this class, I will add a unit test for this method now, and open a JIRA https://issues.cask.co/browse/CDAP-15558 to track other methods.